syllabus/task_space/task_space.py

```diff
@@ -11,34 +11,179 @@ def __init__(self, gym_space: Union[Space, int], tasks=None):
+            
+            combinations = [p for p in itertools.product(*tasks)]
+            encode_map = {task: i for i, task in enumerate(combinations)}
+            decode_map = {i: task for i, task in enumerate(combinations)}
+    
+            encoder = lambda task: encode_map[task] if task in encode_map else None
+            decoder = lambda task: decode_map[task] if task in decode_map else None
+
+        elif isinstance(space, Dict):
+
+            # Function flatten() takes a nested dictionary as input and flattens it to have each key in
+            # root's level. Keys whos values are dictionaries store the length of keys. 
+            # This function will help create a mapping dictionary that can store the encoder that will use the 
+            # space and tasks value based on keys.
+
+            def flatten(d,out):
+                for (key,value) in d.items() :
+                    if ( isinstance(value, dict) or isinstance(value, Dict))  :
+                        out.update({key : len(value)})
+                        out.update(flatten(value, {}))
+                    else:
+                        out[key] = value
+                return out
+
+            # Flatten spaces and tasks
+
+            # Example :  
+
+            # SPACE : gym.spaces.Dict(  
+                        # {
+                        #     "ext_controller": gym.spaces.MultiDiscrete([5, 2, 2]),
+                        #     "inner_state": gym.spaces.Dict(
+                        #         {
+                        #             "charge": gym.spaces.Discrete(10),
+                        #             "system_checks": gym.spaces.Tuple((gym.spaces.MultiDiscrete([3,2]),gym.spaces.Discrete(3))),
+                        #             "job_status": gym.spaces.Dict(
+                        #                 {
+                        #                     "task": gym.spaces.Discrete(5),
+                        #                     "progress": gym.spaces.Box(low=0, high=1, shape=(2,0)),
+                        #                 }
+                        #             ),
+                        #         }
+                        #     ),
+                        # }
+                        # )
+            space = flatten(space.spaces ,{})
+
+            # TASK :  {
+                    #     "ext_controller": [("a", "b", "c"),(1,0),("X","Y")],
+                    #     "inner_state": {
+                    #             "charge": [0,1,2,3,4,5,6,7,8,9],
+                    #             "system_checks": ((("a", "b", "c"),(1,0)),("X","Y","Z")),
+                    #             "job_status": {
+                    #                     "task": ["A","B", "C", "D", "E"],
+                    #                     "progress": [(0, 0), (0, 1), (1, 0), (1, 1)],
+                    #                 }
+                            
+                    #         }
+                    # }
+
```

Comment:
Please delete any commented out code. If you still need it, move it to a text file somewhere